strawberryfields/program.py

```diff
@@ -500,23 +552,27 @@ def compile(self, target, **kwargs):
         if kwargs.get("optimize", False):
             seq = pu.optimize_circuit(seq)
 
-        # does the circuit spec  have its own compilation method?
-        if db.compile is not None:
-            seq = db.compile(seq, self.register)
+        seq = compiler.compile(seq, self.register)
 
         # create the compiled Program
         compiled = self._linked_copy()
         compiled.circuit = seq
-        compiled._target = db.short_name
+        compiled._target = target
+
+        # Get run options of compiled program.
+        run_options = {k: kwargs[k] for k in ALLOWED_RUN_OPTIONS if k in kwargs}
+        compiled.run_options.update(run_options)
 
-        # get run options of compiled program
-        # for the moment, shots is the only supported run option.
-        if "shots" in kwargs:
-            compiled.run_options["shots"] = kwargs["shots"]
+        # set backend options of the program
+        backend_options = {k: kwargs[k] for k in kwargs if k not in ALLOWED_RUN_OPTIONS}
+        compiled.backend_options.update(backend_options)
```

Comment:
this allows for non allowed backend options too, just checking if that's alright
```python
import strawberryfields as sf
import strawberryfields.ops as ops

prog = sf.Program(8)

with prog.context as q:
    ops.MeasureFock() | q

eng = sf.RemoteEngine("X8")
device = eng.device_spec
prog2 = prog.compile(device, **{'some_kwarg:': 1})
print(prog2.backend_options)
```
```
{'some_kwarg:': 1}
```

Reply:
Yep, hardware backend options might be unknown by SF, so we can't validate them here! They will be validated by the API instead.

Reply:
I would like to revisit how we pass run and backend options around the place, though. It is very 'ad hoc' and should be cleaned up.